File: krishanaposhak-frontend/src/utils/shippingCalculator.js
import { formatPrice } from './formatPrice';

export const FREE_SHIPPING_THRESHOLD = 8000;

/**
 * Calculates shipping charge based on order subtotal according to business rules:
 * - Subtotal < ₹2,000                => Shipping = ₹120
 * - Subtotal ₹2,000 – ₹3,999.99     => Shipping = ₹240
 * - Subtotal ₹4,000 – ₹7,999.99     => Shipping = ₹400
 * - Subtotal ≥ ₹8,000               => Shipping = ₹0 (FREE DELIVERY)
 *
 * @param {number} subtotal - The cart/order subtotal amount in INR
 * @returns {{
 *   shipping: number,
 *   isFreeShipping: boolean,
 *   freeShippingThreshold: number,
 *   remainingForFreeShipping: number,
 *   freeShippingMessage: string
 * }}
 */
export function calculateShipping(subtotal = 0) {
  const numericSubtotal = Math.max(0, Number(subtotal) || 0);

  let shipping = 0;
  if (numericSubtotal < 2000) {
    shipping = 120;
  } else if (numericSubtotal < 4000) {
    shipping = 240;
  } else if (numericSubtotal < 8000) {
    shipping = 400;
  } else {
    shipping = 0;
  }

  const isFreeShipping = shipping === 0;
  const remainingForFreeShipping = isFreeShipping
    ? 0
    : Math.max(0, FREE_SHIPPING_THRESHOLD - numericSubtotal);

  const freeShippingMessage = isFreeShipping
    ? 'FREE DELIVERY'
    : `Add ${formatPrice(remainingForFreeShipping)} more to get FREE Delivery`;

  return {
    shipping,
    isFreeShipping,
    freeShippingThreshold: FREE_SHIPPING_THRESHOLD,
    remainingForFreeShipping,
    freeShippingMessage,
  };
}

export default calculateShipping;

File: krishanaposhak/src/main/java/com/tanuj/krishanaposhak/util/ShippingCalculator.java
package com.tanuj.krishanaposhak.util;

import java.math.BigDecimal;

/**
 * Utility class for calculating shipping charges based on order subtotal according to business rules:
 * - Subtotal == 0                   => Shipping = ₹0
 * - Subtotal < ₹2,000               => Shipping = ₹120
 * - Subtotal ₹2,000 – ₹3,999.99     => Shipping = ₹240
 * - Subtotal ₹4,000 – ₹7,999.99     => Shipping = ₹400
 * - Subtotal ≥ ₹8,000               => Shipping = ₹0 (FREE DELIVERY)
 */
public final class ShippingCalculator {

    public static final double FREE_SHIPPING_THRESHOLD = 8000.0;

    private ShippingCalculator() {
        // Utility class
    }

    /**
     * Calculates shipping charge as a BigDecimal.
     *
     * @param amount Subtotal after discount or cart subtotal
     * @return Shipping charge
     */
    public static BigDecimal calculateShippingCharge(BigDecimal amount) {
        if (amount == null || amount.compareTo(BigDecimal.ZERO) <= 0) {
            return BigDecimal.ZERO;
        }
        double val = amount.doubleValue();
        if (val < 2000.0) {
            return BigDecimal.valueOf(120.0);
        } else if (val < 4000.0) {
            return BigDecimal.valueOf(240.0);
        } else if (val < 8000.0) {
            return BigDecimal.valueOf(400.0);
        } else {
            return BigDecimal.ZERO;
        }
    }

    /**
     * Calculates shipping charge as a double.
     *
     * @param amount Subtotal amount
     * @return Shipping charge
     */
    public static double calculateShippingCharge(double amount) {
        if (amount <= 0.0) {
            return 0.0;
        }
        if (amount < 2000.0) {
            return 120.0;
        } else if (amount < 4000.0) {
            return 240.0;
        } else if (amount < 8000.0) {
            return 400.0;
        } else {
            return 0.0;
        }
    }
}
